### PR TITLE
Introduce MetricRecorder

### PR DESCRIPTION
**Pull request by cbalioglu**
This PR:
- introduces a new `MetricRecorder` interface and `LogMetricRecorder` and `TensorBoardRecorder` classes.
- revises how we format metric values and extracts the relevant functions from `MetricBag`.
- performs a couple nit readability updates in other parts of the code.

### CODE REVIEW

File: src/fairseq2/metrics.py [hunk 1]
```diff
@@ -158,24 +158,218 @@ def sync_and_compute_metrics(*bags: MetricBag) -> Optional[Dict[str, Any]]:
...
+format_as_loss = partial(format_as_float, decimal=3)
+"""Format metric ``value`` as training loss."""
+
+
+_metric_formatters: Dict[str, Tuple[str, Callable[[Any], str]]] = {
+    # fmt: off
+    "batch_size":          ("Batch Size",                format_as_int),
+    "elapsed_time":        ("Elapsed Time",              format_as_seconds),
+    "elements_per_batch":  ("Elements per Batch",        format_as_int),
+    "elements_per_second": ("Elements per Second",       format_as_int),
```
Comment (shagunsodhani) [inline src/fairseq2/metrics.py:201]: this can be a float right ? e.g. if we process 1 element per 400 ms ?

File: src/fairseq2/metrics.py [hunk 2]
```diff
@@ -158,24 +158,218 @@ def sync_and_compute_metrics(*bags: MetricBag) -> Optional[Dict[str, Any]]:
...
+        :param values:
+            The metric values.
+        :param step_nr:
+            The number of the run step.
+        :param flush:
+            If ``True``, flushes any intermediate buffer after recording.
+        """
+
+    def close(self) -> None:
+        """Close the recorder."""
```
Comment (shagunsodhani) [inline src/fairseq2/metrics.py:258]: nit: shouldnt this have a `pass` statement or something like that

File: src/fairseq2/metrics.py (hunk 1 above)
Comment (cbalioglu) [reply src/fairseq2/metrics.py:201]: These formatters are meant to be used to report progress via stdout/stderr. For values less that 1.0, we just show `<1`. For instance, see the elapsed time in this log output:

```
Valid:eng-yor Metrics (step 200) - Loss: 10.645 | Entropy Loss: 15.358 | Batch Size: 248 | Elements per Batch: 11,304 | Elements per Second: 46,781 | Number of Source Elements: 33,447 | Number of Target Elements: 45,217 | Elapsed Time: <1s
```

File: src/fairseq2/metrics.py (hunk 2 above)
Comment (cbalioglu) [reply src/fairseq2/metrics.py:258]: It does not necessarily have to have, syntactically this is correct, but to increase the readability I added a "redundant" `pass` in #287.

File: src/fairseq2/metrics.py (hunk 1 above)
Comment (shagunsodhani) [reply src/fairseq2/metrics.py:201]: Thank you for the clarification!

File: src/fairseq2/metrics.py (hunk 2 above)
Comment (shagunsodhani) [reply src/fairseq2/metrics.py:258]: Thank you for the clarification!

File: src/fairseq2/metrics.py (hunk 1 above)
Comment (cbalioglu) [reply src/fairseq2/metrics.py:201]: No problem!
